ergo-rest/src/api/peer_discovery_internals.rs

```diff
@@ -0,0 +1,410 @@
+                            Ok(_) => {
+                                visited_peers.insert(url);
+                                count += 1;
+                            }
+                            Err(TrySendError::Full) => {
+                                // Push it back on the stack, try again later.
+                                peer_stack.push(peer);
+                                break;
+                            }
+                            Err(TrySendError::Closed) => {
+                                return Err(PeerDiscoveryError::MpscSender);
+                            }
+                        }
+                    }
+                }
+                match p {
+                    Msg::AddActiveNode(mut url) => {
+                        #[allow(clippy::unwrap_used)]
+                        url.set_port(None).unwrap();
+                        //println!("Active node {}", url);
+                        visited_active_peers.insert(url.clone());
+                        visited_peers.insert(url);
+                        count -= 1;
+                        if count == 0 {
+                            break 'loop_;
+                        }
+                    }
+                    Msg::AddInactiveNode(mut url) => {
+                        #[allow(clippy::unwrap_used)]
+                        url.set_port(None).unwrap();
+                        visited_peers.insert(url);
+                        count -= 1;
+                        if count == 0 {
+                            break 'loop_;
+                        }
+                    }
+                    Msg::CheckPeers(peers) => {
+                        peer_stack.extend(peers);
+                    }
+                }
+            }
+            C::RxTimeoutSignal => {
+                break 'loop_;
+            }
+        }
+    }
+
+    //println!("Total # nodes visited: {}", visited_peers.len());
+
+    drop(tx_url);
+    let coll: Vec<_> = visited_active_peers
+        .difference(&seeds_set)
+        .into_iter()
+        .cloned()
+        .collect();
+    Ok(coll)
+}
+
+/// Given a stream that receives URLs of full ergo nodes, spawn a task (task 2 in the schematic
+/// above) which checks if it is active.  If so, request its peers. In all cases, a message (enum
+/// `Msg`) is sent out to notify the listener.
+fn spawn_http_request_task<
+    SendMsg: ChannelInfallibleSender<Msg> + Clone + Send + Sync + 'static,
+>(
+    tx_peer: SendMsg,
+    url_stream: impl futures::Stream<Item = Url> + Send + 'static,
+    max_parallel_requests: BoundedU16<1, { u16::MAX }>,
+    request_timeout_duration: Duration,
+) {
+    use futures::StreamExt;
+
+    // Note that `tokio` - the de facto standard async runtime - is not supported on WASM. We need
+    // to spawn tasks for HTTP requests, and for WASM we rely on the `wasm_bindgen_futures` crate.
+    #[cfg(not(target_arch = "wasm32"))]
+    let spawn_fn = tokio::spawn;
```

Comment:
Wasn't able to abstract out the spawn functions with traits (just couldn't type check it). Used conditional compilation instead. It's ugly but it works.